### PR TITLE
I5pranay93/infws 161 listing apis

### PR DESCRIPTION
**Pull request by i5pranay93**
## Description
Contains 4 set of API's

1. Linking of the repository to the app.
2. Listing of the repository
3. listing of the cookbooks(specific repository and overall)
4. listing of recipes and it's count

Public Link of API's call - https://www.postman.com/collections/3dd7a72730d0b2571078

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: src/workstation-gui/app/controllers/api/v1/repositories_controller.rb
Comment (vkarve-chef) [inline src/workstation-gui/app/controllers/api/v1/repositories_controller.rb:80]: perhaps just use [split](https://ruby-doc.org/core-3.1.2/File.html#method-c-split)?

File: src/workstation-gui/app/helpers/application_helper.rb
Comment (vkarve-chef) [inline src/workstation-gui/app/helpers/application_helper.rb:33]: should this be `validate_dir_path`?

File: src/workstation-gui/app/helpers/application_helper.rb
Comment (vkarve-chef) [inline src/workstation-gui/app/helpers/application_helper.rb:55]: the `'json'` filter could be excluded as that will get us the `<custom name policy file>.lock.json`.

**Comment by vkarve-chef** [comment]
Good work on tests added 👏

**Review by vkarve-chef — APPROVED** [review]

File: src/workstation-gui/Gemfile
Comment (sanjain-progress) [inline src/workstation-gui/Gemfile:87]: as well as the 'rspec-rails' gem in group:test,:development. We don't require this gem when we ship the workstation-app.

File: src/workstation-gui/app/controllers/api/v1/cookbooks_controller.rb
Comment (sanjain-progress) [inline src/workstation-gui/app/controllers/api/v1/cookbooks_controller.rb:28]: Is it okay to have `success` as Constant and use it from there. So we have a single source. I will also do the same in this [PR](https://github.com/chef/chef-workstation/pull/2926/files)

File: src/workstation-gui/app/controllers/api/v1/cookbooks_controller.rb
Comment (sanjain-progress) [inline src/workstation-gui/app/controllers/api/v1/cookbooks_controller.rb:45]: why we this `validate_creds` here

File: src/workstation-gui/app/controllers/api/v1/recipes_controller.rb
Comment (sanjain-progress) [inline src/workstation-gui/app/controllers/api/v1/recipes_controller.rb:38]: same here `validate_creds`

File: src/workstation-gui/app/controllers/api/v1/repositories_controller.rb
Comment (sanjain-progress) [inline src/workstation-gui/app/controllers/api/v1/repositories_controller.rb:96]: Can we take this as an improvement in the next cycle to follow the design pattern like a [Skinny controller and fat model](https://riptutorial.com/ruby-on-rails/example/9609/fat-model--skinny-controller)? As in MVC, business logic mainly goes into the model.

File: src/workstation-gui/app/helpers/application_helper.rb
Comment (sanjain-progress) [inline src/workstation-gui/app/helpers/application_helper.rb:22]: I see you include this module in `ApplicationController` and there are a lot of methods is written here. 

So this can become a performance issue because every controller inherits from this controller. As this method will always be in memory irrespective controller is invoking that method or not. Usually, we write only those which are needed for every request like Authentication code, pagination

File: src/workstation-gui/spec/requests/api/v1/cookbooks_spec.rb
Comment (sanjain-progress) [inline src/workstation-gui/spec/requests/api/v1/cookbooks_spec.rb:1]: Looks really good, you wrote test cases for your code.

**Review by sanjain-progress — APPROVED** [review]

File: src/workstation-gui/app/controllers/api/v1/repositories_controller.rb
Comment (i5pranay93) [reply src/workstation-gui/app/controllers/api/v1/repositories_controller.rb:80]: it Split by '/' and takes last element (which is cookbook name)

File: src/workstation-gui/app/helpers/application_helper.rb
Comment (i5pranay93) [reply src/workstation-gui/app/helpers/application_helper.rb:33]: renamed it.

File: src/workstation-gui/Gemfile
Comment (i5pranay93) [reply src/workstation-gui/Gemfile:57]: sure, but we don't have anything like a development and production environment so I kept it outside only. i will remove it as you have already added it.

File: src/workstation-gui/Gemfile
Comment (i5pranay93) [reply src/workstation-gui/Gemfile:87]: This one is in group test only.

File: src/workstation-gui/app/controllers/api/v1/cookbooks_controller.rb
Comment (i5pranay93) [reply src/workstation-gui/app/controllers/api/v1/cookbooks_controller.rb:28]: Once you do it and we merge, it i can change it ( i am waiting for merge) as it will let me add services and use some of yours model (for e.g if you made service folder and i did that too that will be duplication and conflict, so some chnages will be optimized after merge)

File: src/workstation-gui/app/controllers/api/v1/cookbooks_controller.rb
Comment (i5pranay93) [reply src/workstation-gui/app/controllers/api/v1/cookbooks_controller.rb:45]: It was added by @ashiqueps  for authentication, i am not using it for now but it has to be used (waiting for merge) and then we can test it along with authentication.

File: src/workstation-gui/app/controllers/api/v1/repositories_controller.rb
Comment (i5pranay93) [reply src/workstation-gui/app/controllers/api/v1/repositories_controller.rb:96]: I have added a todo to move them to services (so once our branch is merged, i will update it)

File: src/workstation-gui/app/helpers/application_helper.rb
Comment (i5pranay93) [reply src/workstation-gui/app/helpers/application_helper.rb:22]: Removed this and moved it into another module helper( WorkstationHelper)

**Comment by vkarve-chef** [comment]
closing old PRs. please reopen as needed.

**Pull request closed without merging**
